### PR TITLE
Feature/add ufs s2s support

### LINKED ISSUE #52: Add Dockerfiles and stack.yamls for UFS containers on AWS
NOAA PSL is exploring the use of hpc-stack to enable UFS on AWS with the use of Docker and Singularity containers
In associated PR's, Docker Doxyfiles for Intel using Intel OneAPI will be added/contributed.
Thanks to @slawrence87544

### PR DESCRIPTION
**Pull request by slawrence87544**
Adding UFS support for AWS
fixes #52

### CODE REVIEW

File: config/config_custom.sh [hunk 1]
```diff
@@ -16,3 +16,4 @@ export   MAKE_CLEAN=N
 export DOWNLOAD_ONLY=N
 export STACK_EXIT_ON_FAIL=Y
 export WGET="wget -nv"
+
```
Comment (aerorahul) [inline config/config_custom.sh]: un-necessary

File: config/stack_ufs_s2s.yaml [hunk 2]
```diff
@@ -0,0 +1,242 @@
+cmake:
```
Comment (aerorahul) [inline config/stack_ufs_s2s.yaml]: Do we need `stack_ufs_s2s.yaml` and `stack_ufs_s2s_model.yaml` both?
How do they differ from `stack_noaa.yaml` or `stack_custom.yaml`.

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    curl -f -s -S -R -L -O https://github.com/Kitware/CMake/releases/download/v3.16.4/cmake-3.16.4-Linux-x86_64.tar.gz && \
+    mkdir -p /usr/local/cmake && \
+    tar zxf cmake-3.16.4-Linux-x86_64.tar.gz -C /usr/local/cmake --strip-components=1 && \
+    rm cmake-3.16.4-Linux-x86_64.tar.gz && \
+    echo export PATH=/usr/local/cmake/bin:\$PATH >> /etc/bashrc
+
+ENV PNETCDF=/usr/local
+ENV NETCDF=/usr/local
+
+RUN git clone https://github.com/slawrence87544/hpc-stack.git && \
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20]: this would become `github.com/noaa-emc`

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    cd hpc-stack && \
+    git checkout feature/add-ufs-s2s-support && \
+    ./build_stack.sh -p /usr/local -c config/config_linux_ubuntu20_intel_oneapi_beta09.sh -y config/stack_ufs_s2s_model.yaml
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20]: Can we build this as `builder` instead in `/usr/local`?

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+RUN cd /tmp && \
+    curl -f -s -S -R -L -O https://github.com/Kitware/CMake/releases/download/v3.16.4/cmake-3.16.4-Linux-x86_64.tar.gz && \
+    mkdir -p /usr/local/cmake && \
+    tar zxf cmake-3.16.4-Linux-x86_64.tar.gz -C /usr/local/cmake --strip-components=1 && \
+    rm cmake-3.16.4-Linux-x86_64.tar.gz && \
+    echo export PATH=/usr/local/cmake/bin:\$PATH >> /etc/bashrc
+
+ENV PNETCDF=/usr/local
+ENV NETCDF=/usr/local
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20]: why are these necessary here?

File: containers/how-to-create-singularity-image-from-local-docker-image.txt [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Comment (aerorahul) [inline containers/how-to-create-singularity-image-from-local-docker-image.txt]: If this is a `.txt` why the `bash` interpreter?   Also, this reads more like a script than a how-to.

File: containers/Dockerfile.intel-oneapi-ufs-s2s-coupled-lite [hunk 7]
```diff
@@ -0,0 +1,14 @@
+FROM intel-oneapi-ufs-s2s-coupled-dev:ubuntu20 as ufs
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-ufs-s2s-coupled-lite]: what is the use case of this dockerfile?

File: libs/build_pio.sh [hunk 8]
```diff
@@ -32,6 +32,7 @@ if $MODULES; then
     fi
 else
     prefix=${PIO_ROOT:-"/usr/local"}
+    PNETCDF=${PNETCDF:-"/usr/local"}
```
Comment (aerorahul) [inline libs/build_pio.sh]: This is not correct.

**Review by aerorahul — CHANGES_REQUESTED** [review]
little comments here and there.

File: config/config_custom.sh (hunk 1 above)
Comment (slawrence87544) [reply config/config_custom.sh]: Sorry, the extra line was unintentional.

File: config/stack_ufs_s2s.yaml (hunk 2 above)
Comment (slawrence87544) [reply config/stack_ufs_s2s.yaml]: I had thought I had removed stack_ufs_s2s.yaml.  The main difference between stack_noaa.yaml and the stack_ufs_s2s_model.yaml are that it doesn't build eigen, ecbuild, eckit, fckit, or atlas as none of those are required.  I was trying to minimize the size of the subsequent container.

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 (hunk 3 above)
Comment (slawrence87544) [reply containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20]: Yes, I'll change that now.  It was only in there so that if someone wanted to try this out prior to the PR being accepted, they could actually do so.  Did you have a preference on how to handle that or is it assumed that people won't try to use this until the changes are merged?

File: libs/build_pio.sh (hunk 8 above)
Comment (slawrence87544) [reply libs/build_pio.sh]: I encountered an issue when building the stack with pnetcdf enabled.  I had PNETCDF defined outside of this script, but when I ran the script, it blew up complaining about an undefined PNETCDF.  Maybe it was user error.  I'll just remove it.

File: containers/Dockerfile.intel-oneapi-ufs-s2s-coupled-lite (hunk 7 above)
Comment (slawrence87544) [reply containers/Dockerfile.intel-oneapi-ufs-s2s-coupled-lite]: In order to reduce the size of the containers, it's common to delete all the source code, but if one simply removes it in the same container image where it was created, it doesn't reduce the size (hence the need for a multi-stage container build).  It really depends on how one plans to use the container, but a smaller container is easier to move around in general and loads up faster when bringing up a cloud-based master instance.

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 (hunk 5 above)
Comment (slawrence87544) [reply containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20]: They may not be necessary, but I was having trouble building some of the packages without them.  Perhaps there is a better way to ensure that those environment variables get created and exported along with NETCDF_ROOT/PNETCDF_ROOT.

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+RUN git clone https://github.com/slawrence87544/hpc-stack.git && \
+    cd hpc-stack && \
+    git checkout feature/add-ufs-s2s-support && \
+    ./build_stack.sh -p /usr/local -c config/config_linux_ubuntu20_intel_oneapi_beta09.sh -y config/stack_ufs_s2s_model.yaml
+
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:45]: Please `rm -rf hpc-stack` after building to lighten up the container. There is no need to retain the stack and build directories of the software.

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 [hunk 10]
```diff
@@ -24,22 +24,19 @@ ENV I_MPI_LIBRARY_KIND='release_mt'
 
 RUN cd /tmp && \
     curl -f -s -S -R -L -O https://github.com/Kitware/CMake/releases/download/v3.16.4/cmake-3.16.4-Linux-x86_64.tar.gz && \
-    mkdir -p /usr/local/cmake && \
-    tar zxf cmake-3.16.4-Linux-x86_64.tar.gz -C /usr/local/cmake --strip-components=1 && \
+    mkdir -p /usr/builder/cmake && \
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20]: I think it is `/home/builder` not `/usr/builder`

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 [hunk 11]
```diff
@@ -24,22 +24,19 @@ ENV I_MPI_LIBRARY_KIND='release_mt'
 
 RUN cd /tmp && \
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:30]: Make sure to delete stuff from `/tmp` to decrease the image size.

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 [hunk 12]
```diff
@@ -24,22 +24,19 @@ ENV I_MPI_LIBRARY_KIND='release_mt'
...
 RUN git clone https://github.com/slawrence87544/hpc-stack.git && \
     cd hpc-stack && \
     git checkout feature/add-ufs-s2s-support && \
-    ./build_stack.sh -p /usr/local -c config/config_linux_ubuntu20_intel_oneapi_beta09.sh -y config/stack_ufs_s2s_model.yaml
+    ./build_stack.sh -p /usr/builder -c config/config_linux_ubuntu20_intel_oneapi_beta09.sh -y config/stack_ufs_s2s_model.yaml
 
-RUN echo 'export PATH=/usr/local/bin:$PATH'      >> /etc/bash.bashrc && \
-    echo 'export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH' >> /etc/bash.bashrc && \
-    echo 'export CMAKE_PREFIX_PATH=/usr/local'   >> /etc/bash.bashrc && \
+RUN echo 'export PATH=/usr/builder/bin:$PATH'      >> /etc/bash.bashrc && \
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20]: The user is `builder` and the space is `/home/builder`

**Review by aerorahul — CHANGES_REQUESTED** [review]
The user `builder` path is `/home/builder` not `/usr/builder`

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 [hunk 13]
```diff
@@ -24,19 +24,19 @@ ENV I_MPI_LIBRARY_KIND='release_mt'
...
+    tar zxf cmake-3.16.4-Linux-x86_64.tar.gz -C /home/builder/cmake --strip-components=1 && \
     rm cmake-3.16.4-Linux-x86_64.tar.gz && \
-    echo export PATH=/usr/builder/cmake/bin:\$PATH >> /etc/bashrc
+    echo export PATH=/home/builder/cmake/bin:\$PATH >> /etc/bashrc
 
 RUN git clone https://github.com/slawrence87544/hpc-stack.git && \
     cd hpc-stack && \
     git checkout feature/add-ufs-s2s-support && \
-    ./build_stack.sh -p /usr/builder -c config/config_linux_ubuntu20_intel_oneapi_beta09.sh -y config/stack_ufs_s2s_model.yaml
+    ./build_stack.sh -p /home/builder -c config/config_linux_ubuntu20_intel_oneapi_beta09.sh -y config/stack_ufs_s2s_model.yaml
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20]: Can you make the install prefix `/home/builder/opt` or something like that?
It is probably not a good idea to treat the home directory as an installation prefix.

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 [hunk 14]
```diff
@@ -24,19 +24,19 @@ ENV I_MPI_LIBRARY_KIND='release_mt'
 
 RUN cd /tmp && \
     curl -f -s -S -R -L -O https://github.com/Kitware/CMake/releases/download/v3.16.4/cmake-3.16.4-Linux-x86_64.tar.gz && \
-    mkdir -p /usr/builder/cmake && \
-    tar zxf cmake-3.16.4-Linux-x86_64.tar.gz -C /usr/builder/cmake --strip-components=1 && \
+    mkdir -p /home/builder/cmake && \
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20]: `/home/builder/opt/cmake`

**Comment by edwardhartnett** [comment]
Is there an issue for this work? If so, please refer to it in the PR. If not, please create one. ;-)

File: containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20 (hunk 9 above)
Comment (slawrence87544) [reply containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:45]: Got it.  I'll remove the src code.  I didn't know if the esmf library needed to preserve it's source code in order to be functional.

**Comment by edwardhartnett** [comment]
@slawrence87544 is there an issue for this work? If not please add one. If so, please reference it in your PR description. 

Also please always add a PR description for any PRs to this repo.

File: containers/Dockerfile.intel-oneapi-ufs-s2s-coupled-dev [hunk 15]
```diff
@@ -0,0 +1,35 @@
+FROM intel-oneapi-hpc-stack:ubuntu20
+
+SHELL ["/bin/bash", "-c"]
+
+RUN apt-get update -y && \
+    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends -o=Dpkg::Use-Pty=0 \
+        csh \
```
Comment (aerorahul) [inline containers/Dockerfile.intel-oneapi-ufs-s2s-coupled-dev]: Do we need all these shells? `csh`, `ksh`, `tcsh`?
Also, why have `tex, lynx, nedit, emacs`? Is `vi` not enough?
We are not building documentation, so `doxygen` is not necessary either.

File: containers/how-to-create-singularity-image-from-local-docker-image.txt [hunk 16]
```diff
@@ -0,0 +1,139 @@
+# This guide assumes that docker is already installed
```
Comment (aerorahul) [inline containers/how-to-create-singularity-image-from-local-docker-image.txt]: Can you make this a markdown.  The formatting is so much better.  I can help.

**Review by aerorahul — CHANGES_REQUESTED** [review]
some minor comments.
getting closer to finished.

File: config/stack_ufs_s2s_model.yaml [hunk 17]
```diff
@@ -0,0 +1,251 @@
+cmake:
```
Comment (aerorahul) [inline config/stack_ufs_s2s_model.yaml]: I can check, but how different is this from `stack_noaa.yaml`

File: containers/Dockerfile.intel-oneapi-ufs-s2s-coupled-dev (hunk 15 above)
Comment (slawrence87544) [reply containers/Dockerfile.intel-oneapi-ufs-s2s-coupled-dev]: These extra packages were included based on a jcsda container found here (https://github.com/JCSDA-internal/containers/blob/develop/Dockerfile.intel19-impi-dev).  Since I don't have much experience with how these dev containers will be used, I put in more shell options as I guessed that researchers might want them to be available.  I will remove them as they are not critical to the ufs-s2s-model's functionality.

File: config/stack_ufs_s2s_model.yaml (hunk 17 above)
Comment (slawrence87544) [reply config/stack_ufs_s2s_model.yaml]: The only differences are inclusion of the jpeg compile and the inclusion of szip functionality for the hdf5 compile.  I can remove them and see how the compile goes.  If it still works, I'll simply remove that config.

Comment (aerorahul) [reply config/stack_ufs_s2s_model.yaml]: jpeg is OFF in stack_noaa.yaml, but we intend to turn it ON.
The UFS uses zlib for compression and that is why the szip is OFF.

**Review by aerorahul — APPROVED** [review]
Thanks @slawrence87544
